File: app/src/main/java/com/github/dedis/student20_pop/model/network/level/high/rollcall/OpenRollCall.java
package com.github.dedis.student20_pop.model.network.level.high.rollcall;

import com.github.dedis.student20_pop.model.network.level.high.Action;
import com.github.dedis.student20_pop.model.network.level.high.Data;
import com.github.dedis.student20_pop.model.network.level.high.Objects;

/**
 * Data sent to open a roll call
 */
public class OpenRollCall extends Data {

    private final String id;
    private final long start;

    public OpenRollCall(String id, long start) {
        this.id = id;
        this.start = start;
    }

    @Override
    public String getObject() {
        return Objects.ROLL_CALL.getObject();
    }

    @Override
    public String getAction() {
        return Action.OPEN.getAction();
    }

    public String getId() {
        return id;
    }

    public long getStart() {
        return start;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        OpenRollCall that = (OpenRollCall) o;
        return getStart() == that.getStart() &&
                java.util.Objects.equals(getId(), that.getId());
    }

    @Override
    public int hashCode() {
        return java.util.Objects.hash(getId(), getStart());
    }

    @Override
    public String toString() {
        return "OpenRollCall{" +
                "id='" + id + '\'' +
                ", start=" + start +
                '}';
    }
}

File: fe1-web/src/features/rollCall/network/messages/OpenRollCall.ts
import { ActionType, MessageData, ObjectType } from 'core/network/jsonrpc/messages';
import { validateDataObject } from 'core/network/validation';
import { checkTimestampStaleness } from 'core/network/validation/Checker';
import { EventTags, Hash, ProtocolError, Timestamp } from 'core/objects';

const paramError = (o: OpenRollCall) => `parameter encountered during roll call ${o.action}`;

/** Data sent to open a Roll-Call event */
export class OpenRollCall implements MessageData {
  public readonly object: ObjectType = ObjectType.ROLL_CALL;

  public readonly action: ActionType = ActionType.OPEN;

  public readonly update_id: Hash;

  public readonly opens: Hash;

  public readonly opened_at: Timestamp;

  constructor(msg: Partial<OpenRollCall>, laoId: Hash) {
    if (!msg.opened_at) {
      throw new ProtocolError(`Undefined 'opened_at' ${paramError(this)}`);
    }
    checkTimestampStaleness(msg.opened_at);
    this.opened_at = msg.opened_at;

    if (!msg.opens) {
      throw new ProtocolError(`Undefined 'opens' ${paramError(this)}`);
    }
    this.opens = msg.opens;

    if (!msg.update_id) {
      throw new ProtocolError(`Undefined 'update_id' ${paramError(this)}`);
    }

    const expectedId = OpenRollCall.computeOpenRollCallId(laoId, this.opens, this.opened_at);

    if (!expectedId.equals(msg.update_id)) {
      throw new ProtocolError(
        `Invalid 'update_id' ${paramError(this)}:` +
          ' re-computing the value yields a different result (' +
          `(expected: '${expectedId}', actual: '${msg.update_id}')`,
      );
    }
    this.update_id = msg.update_id;
  }

  /**
   * Computes the id for an opened roll call
   * @param laoId The id of the lao this roll call takes place in
   * @param rollCallIdToOpen The id of the roll call that is opened with this message
   * @param openedAt The time the roll call is opened
   * @returns The id for the opened roll call
   */
  public static computeOpenRollCallId(
    laoId: Hash,
    rollCallIdToOpen: Hash,
    openedAt: Timestamp,
  ): Hash {
    return Hash.fromStringArray(
      EventTags.ROLL_CALL,
      laoId.valueOf(),
      rollCallIdToOpen.valueOf(),
      openedAt.toString(),
    );
  }

  /**
   * Creates an OpenRollCall object from a given object.
   *
   * @param obj The parsed json data
   * @param laoId The id of the lao this roll call is in
   */
  public static fromJson(obj: any, laoId?: Hash): OpenRollCall {
    if (!laoId) {
      throw new Error(
        "Tried build a 'ReopenRollCall' message without knowing the associated lao id",
      );
    }

    const { errors } = validateDataObject(ObjectType.ROLL_CALL, ActionType.OPEN, obj);

    if (errors !== null) {
      throw new ProtocolError(`Invalid open roll call\n\n${errors}`);
    }

    return new OpenRollCall(
      {
        ...obj,
        opened_at: new Timestamp(obj.opened_at),
        update_id: new Hash(obj.update_id),
        opens: new Hash(obj.opens),
      },
      laoId,
    );
  }
}
